README.rst
=============
WSGI Harakiri
=============

.. image:: https://img.shields.io/pypi/v/wsgi-harakiri.svg
        :target: https://pypi.python.org/pypi/wsgi-harakiri

.. image:: https://img.shields.io/travis/adamchainz/wsgi-harakiri.svg
        :target: https://travis-ci.org/adamchainz/wsgi-harakiri

WSGI Middleware that implements a customizable 'harakiri' like uWSGI.


Installation
============

Use **pip**:

.. code-block:: bash

    pip install wsgi-harakiri

Tested on Python 2.7, 3.4, and 3.5.

Usage
=====

Wrap your WSGI application with the middleware, for example for a Django
application in your ``wsgi.py``:

.. code-block:: python

    import os

    from django.core.wsgi import get_wsgi_application

    os.environ.setdefault("DJANGO_SETTINGS_MODULE", "myproject.settings")

    application = get_wsgi_application()

    from wsgi_harakiri import HarakiriMiddleware

    # By default adds a request timeout of 30 seconds
    application = HarakiriMiddleware(application)

Your app **must not** be running a multi-threaded server (multi-process is ok),
and it **must** be running on a POSIX system. The ``alarm`` system call is
used, so this **cannot** be combined with other things that use it, e.g. the
'harakiri' functionality in uWSGI.

API
===

``HarakiriMiddleware(application, timeout=30, on_harakiri=None, error_app=None)``
---------------------------------------------------------------------------------

Wraps a WSGI application with the harakiri functionality.

``application`` may be any WSGI application.

``timeout`` may be any integer number of seconds, and defaults to 30.

``on_harakiri`` will be called when a harakiri occurs, from inside the alarm
function - it is thus suitable for logging the stack trace that lead to the
timeout. Its return value is ignored.

``error_app`` is a second WSGI application that will be called to produce an
error response when a timeout occurs. The default response is a rather plain
'500 Internal server error' with HTML '<h1>Page load timeout</h1>'.

Example usage with all arguments:

.. code-block:: python

    application = get_wsgi_application()

    from wsgi_harakiri import HarakiriMiddleware


    def harakiri_handler():
        logger.error("Harakiri occured", extra={'stack': True})


    def harakiri_page(environ, start_response):
        start_response(
            '500 Internal server error',
            [('Content-Type', 'text/html')]
        )
        return [b'<h1>Sorry, this page timed out.</h1>']


    application = HarakiriMiddleware(
        application,
        timeout=15,
        on_harakiri=harakiri_handler,
        error_app=harakiri_page,
    )

``Harakiri``
------------

This is the exception that gets raised when a timeout occurs. You should
**not** catch it anywhere in your code, however you could use it to detect when
it happens inside a particular code path. For example:

.. code-block:: python

    from wsgi_harakiri import Harakiri


    def find_users(search_term):
        conn = make_db_connection()
        try:
            return conn.query(search_term)
        except Harakiri:
            logger.error("A search timed out", extra={'search_term': search_term})
            raise


HISTORY.rst
=======
History
=======

1.0.0 (2016-03-28)
------------------

* First release on PyPI.


tests/test_it.py
# -*- encoding:utf-8 -*-
from __future__ import absolute_import, division, print_function, unicode_literals

import signal
import subprocess
import time
from contextlib import contextmanager

import requests

from wsgi_harakiri import HarakiriMiddleware


@contextmanager
def run_gunicorn(app_name):
    command = ['gunicorn', __name__ + ':' + app_name]
    process = subprocess.Popen(command)
    time.sleep(1)  # Let gunicorn start
    try:
        yield
    finally:
        process.send_signal(signal.SIGINT)
        process.wait()


def test_times_out():
    with run_gunicorn('app_times_out'):
        resp = requests.get('http://localhost:8000')
        assert resp.status_code == 500
        assert 'Page load timeout' in resp.content


def app_times_out(environ, start_response):
    time.sleep(1.1)

app_times_out = HarakiriMiddleware(app_times_out, 1)


def test_does_not_time_out():
    with run_gunicorn('app_does_not_time_out'):
        resp = requests.get('http://localhost:8000')
        assert resp.status_code == 200
        assert resp.content == '<h1>Fine</h1>'


def app_does_not_time_out(environ, start_response):
    start_response(
        '200 OK',
        [('Content-Type', 'text/html')]
    )
    return [b'<h1>Fine</h1>']

app_does_not_time_out = HarakiriMiddleware(app_does_not_time_out, 2)


def test_custom_handler():
    with run_gunicorn('app_custom_handler'):
        resp = requests.get('http://localhost:8000')
        # Error page from gunicorn itself
        assert resp.status_code == 500
        assert '<h1><p>Internal Server Error</p></h1>' in resp.content


def app_custom_handler(environ, start_response):
    time.sleep(1.1)


def custom_handler():
    raise ValueError()

app_custom_handler = HarakiriMiddleware(app_custom_handler, 1, custom_handler)


def test_custom_error():
    with run_gunicorn('app_custom_error'):
        resp = requests.get('http://localhost:8000')
        assert resp.status_code == 418
        assert resp.content == "I'm a teapot"


def app_custom_error(environ, start_response):
    time.sleep(1.1)


def custom_error(environ, start_response):
    start_response(
        "418 I'm a teapot",
        [('Content-Type', 'text/plain')]
    )
    return [b"I'm a teapot"]


app_custom_error = HarakiriMiddleware(app_custom_error, 1, error_app=custom_error)


runtests.py
#!/usr/bin/env python
# -*- coding: utf-8 -*-
from __future__ import print_function

import sys
import subprocess

import pytest


def main():
    try:
        sys.argv.remove('--nolint')
    except ValueError:
        run_lint = True
    else:
        run_lint = False

    try:
        sys.argv.remove('--lintonly')
    except ValueError:
        run_tests = True
    else:
        run_tests = False

    if run_tests:
        exit_on_failure(tests_main())

    if run_lint:
        exit_on_failure(run_flake8())
        exit_on_failure(run_isort())

        # Broken on 2.7.<9 and 3.4.<3 due to http://bugs.python.org/issue23063
        if not (
            (sys.version_info[:2] == (2, 7) and sys.version_info[2] <= 9) or
            (sys.version_info[:2] == (3, 4) and sys.version_info[2] <= 2)
        ):
            exit_on_failure(run_setup_py_check())


def tests_main():
    return pytest.main()


def run_flake8():
    print('Running flake8 code linting')
    ret = subprocess.call([
        'flake8',
        '--exclude=.svn,CVS,.bzr,.hg,.git,__pycache__,.tox',
        'wsgi_harakiri', 'tests'
    ])
    print('flake8 failed' if ret else 'flake8 passed')
    return ret


def run_isort():
    print('Running isort check')
    return subprocess.call([
        'isort', '--recursive', '--check-only', '--diff',
        '-a', 'from __future__ import absolute_import, division',
        '-a', 'from __future__ import print_function, unicode_literals',
        'wsgi_harakiri', 'tests'
    ])


def run_setup_py_check():
    print('Running setup.py check')
    return subprocess.call([
        'python', 'setup.py', 'check',
        '-s', '--restructuredtext', '--metadata'
    ])


def exit_on_failure(ret, message=None):
    if ret:
        sys.exit(ret)


if __name__ == '__main__':
    main()


wsgi_harakiri/__init__.py
# -*- encoding:utf-8 -*-
from __future__ import absolute_import, division, print_function, unicode_literals

import signal

import six


class HarakiriMiddleware(object):

    def __init__(self, app, timeout=30, on_harakiri=None, error_app=None):
        self.app = app

        if not isinstance(timeout, six.integer_types):
            raise ValueError("Timeout must be an integer.")
        if on_harakiri is not None:
            if not callable(on_harakiri):
                raise ValueError("on_harakiri must be a callable.")
        self.context = HarakiriContext(timeout, on_harakiri)

        if error_app is not None:
            if not callable(error_app):
                raise ValueError("error_app must be a callable.")
            self.error_app = error_app
        else:
            self.error_app = simple_timeout_app

    def __call__(self, environ, start_response):
        try:
            with self.context:
                return self.app(environ, start_response)
        except Harakiri:
            return self.error_app(environ, start_response)


def simple_timeout_app(environ, start_response):
    start_response(
        '500 Internal server error',
        [('Content-Type', 'text/html')]
    )
    return [b'<h1>Page load timeout</h1>']


class HarakiriContext(object):
    def __init__(self, timeout, on_harakiri):
        self.timeout = timeout
        self.on_harakiri = on_harakiri

    def __enter__(self):
        signal.signal(signal.SIGALRM, self.raise_harakiri)
        signal.alarm(self.timeout)

    def __exit__(self, _, __, ___):
        signal.alarm(0)

    def raise_harakiri(self, signum, frame):
        if self.on_harakiri is not None:
            self.on_harakiri()
        raise Harakiri()


class Harakiri(Exception):
    pass
